src/audio/mixin_mixout/mixin_mixout_hifi5.c

```diff
@@ -0,0 +1,285 @@
+		 * one by one to avoid memory access overrun
+		 */
+		for (i = 0; i < left ; i++) {
+			AE_L16_IP(in_sample, (ae_int16 *)in, sizeof(ae_int16));
+			AE_L16_IP(out_sample, (ae_int16 *)out, 0);
+			out_sample = AE_ADD16S(in_sample, out_sample);
+			AE_S16_0_IP(out_sample, (ae_int16 *)out, sizeof(ae_int16));
+		}
+	}
+
+	for (left_samples = samples_to_copy; left_samples > 0; left_samples -= n) {
+		src = cir_buf_wrap(src + n, source->buf_start, source->buf_end);
+		dst = cir_buf_wrap(dst + n, sink->buf_start, sink->buf_end);
+		/* calculate the remaining samples*/
+		nmax = (ae_int16 *)source->buf_end - src;
+		n = AE_MIN32(left_samples, nmax);
+		nmax = (ae_int16 *)sink->buf_end - dst;
+		n = AE_MIN32(n, nmax);
+		in = (ae_int16x8 *)src;
+		out = (ae_int16x8 *)dst;
+		inu = AE_LA128_PP(in);
+		m = n >> 3;
+		left = n & 0x07;
+		/* process 8 frames per loop */
+		for (i = 0; i < m; i++) {
+			AE_LA16X4X2_IP(in_sample, in_sample1, inu, in);
+			AE_SA16X4X2_IP(in_sample, in_sample1, outu2, out);
+		}
+		AE_SA128POS_FP(outu2, out);
+
+		/* process the left samples that less than 8
+		 * one by one to avoid memory access overrun
+		 */
+		for (i = 0; i < left ; i++) {
+			AE_L16_IP(in_sample, (ae_int16 *)in, sizeof(ae_int16));
+			AE_S16_0_IP(in_sample, (ae_int16 *)out, sizeof(ae_int16));
+		}
+	}
+}
+#endif	/* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+static void mix_s24(struct cir_buf_ptr *sink, int32_t start_sample, int32_t mixed_samples,
+		    const struct cir_buf_ptr *source,
+		    int32_t sample_count, uint16_t gain)
+{
+	int samples_to_mix, samples_to_copy, left_samples;
+	int n, nmax, i, m, left;
+	ae_int32x2 in_sample, in_sample1;
+	ae_int32x2 out_sample, out_sample1;
+	ae_int32x4 *in;
+	ae_int32x4 *out;
+	ae_valignx2 inu = AE_ZALIGN128();
+	ae_valignx2 outu1 = AE_ZALIGN128();
+	ae_valignx2 outu2 = AE_ZALIGN128();
+	/* cir_buf_wrap() is required and is done below in a loop */
+	int32_t *dst = (int32_t *)sink->ptr + start_sample;
+	int32_t *src = source->ptr;
+
+	assert(mixed_samples >= start_sample);
+	samples_to_mix = AE_MIN32(mixed_samples - start_sample, sample_count);
+	samples_to_copy = sample_count - samples_to_mix;
+	n = 0;
+
+	for (left_samples = samples_to_mix; left_samples > 0; left_samples -= n) {
+		src = cir_buf_wrap(src + n, source->buf_start, source->buf_end);
+		dst = cir_buf_wrap(dst + n, sink->buf_start, sink->buf_end);
+		/* calculate the remaining samples*/
+		nmax = (int32_t *)source->buf_end - src;
+		n = AE_MIN32(left_samples, nmax);
+		nmax = (int32_t *)sink->buf_end - dst;
+		n = AE_MIN32(n, nmax);
+		in = (ae_int32x4 *)src;
+		out = (ae_int32x4 *)dst;
+		inu = AE_LA128_PP(in);
+		outu1 = AE_LA128_PP(out);
+		m = n >> 2;
+		left = n & 3;
+		/* process 2 samples per time */
+		for (i = 0; i < m; i++) {
+			AE_LA32X2X2_IP(in_sample, in_sample1, inu, in);
+			AE_LA32X2X2_IP(out_sample, out_sample1, outu1, out);
+			out--;
+			out_sample = AE_ADD24S(in_sample, out_sample);
```

Comment:
Has this mix code been been verified? The ALSA S24_LE is aligned to least significant bits while ae_f24 in Xtensa is most significant bits aligned. In the past at least SOF was using the ALSA S24_LE. We used to have a mess with this and I'm no more sure if we have discarded the 24 bit ALSA format to be compatible with Windows OS's 24 bit format. Is the previous generic or HiFi3 mixing like this, if yes this would be OK. Changing 24 bit assumptions is another discussion and apparently we may  support both in parts of code (cc @kv2019i ). But we should reject in prepare() unsupported 24 bit flavor.

Reply:
ok - having a reject message would be fine here until support for both 24bit formats is upstream. This message and 24 bit support can be added incrementally though since this is good for testing now and passing CI.

Reply:
I hope CI is testing the 24 bit case. Proposing to make another test patch that breaks mix_s24() function and see if it fails the CI run.